content/blog/2020-12-20-bril-algebraic-data-types/index.md

```diff
@@ -0,0 +1,538 @@
+this because it is not a good idea, so I do not consider this to be a
+problem.
+
+
+## Implementation: Rust compiler
+
+The Rust compiler is a [Rust
+program](https://github.com/Calsign/bril/blob/algebraic-types/bril-rs/src/main.rs)
+that accepts a Rust source file on standard input and outputs a
+JSON-formatted Bril program on standard output that implements the
+input Rust code. It supports a small but useful subset of the Rust
+language syntax.
+
+The program can be built with `cargo build` from the `bril-rs`
+directory and run with `cargo run` from that directory.
+
+Critically, the compiler makes no attempt to follow the semantics of
+Rust; for example, there is no type checker or borrow checker. The
+compiler just leverages Rust syntax to make it easier to write Bril
+programs.
+
+The compiler is akin to the [TypeScript
+compiler](https://capra.cs.cornell.edu/bril/tools/ts2bril.html). I
+decided to make a new compiler in Rust rather than expanding on the
+existing TypeScript one because I wanted to compile a language that
+had named variant types. (I also wanted to write the compiler in a
+language that I preferred.)
+
+The compiler uses the Rust [syn
+package](https://docs.rs/syn/1.0.54/syn/) to parse the source code.
+
+### Supported Rust features
+
+The Rust compiler supports the following Rust types:
+
+ - `i32`
+ - `bool`
+ - tuples
+ - enum variants with arbitrary payloads
+
+Note that the compiler does not support variants with multiple
+constructors that have the same payload type. For example, here is an
+enum that is not permitted:
+
+```rust
+enum Bad {
+    FirstConstructor(i32),
+    SecondConstructor(i32),
+}
+```
+
+The reason for this restriction is that the underlying Bril
+representation is an anonymous sum type and the interpreter can't tell
+the difference between two constructors unless their payloads have
+different types.
+
+The Rust compiler supports the following Rust syntactic constructs:
+
+ - Functions
+ - Expressions
+   - Constants (`i32`, `bool`)
+   - Unary operators (`-`, `not`)
+   - Arithmetic binary operators (`+`, `-`, `*`, `/`)
+   - Comparison binary operators (`=`, `<`, `<=`, `>`, `>=`)
+   - Logical binary operators (`&&`, `||`)
+   - Tuple creation
+   - Tuple access (e.g., `tuple.0`)
+   - Constructors
+   - If
+   - While
+   - Function calls
+   - Match
+ - Let bindings
+ - Return
+ - `println!` macro
+ - Variant declarations
+
+Expressions can be arbitrarily nested because the compiler performs a
+primitive form of tiling, generating arbitrary temporary variables as
+necessary. The resulting Bril code is very bloated as a result and
+could benefit tremendously from some basic optimization passes.
+
+Let bindings must have type annotations attached, for example:
+
+```rust
+let x: i32 = 42;
+```
+
+This restriction comes from the fact that all Bril value instructions
+require a type. It would be possible to add type inference to the
+compiler, but that would have been a lot more work.
```

Comment:
Yeah. I guess what you'd actually want is to hook into the actual Rust compiler's AST and reuse its type checker.